sdk/storage/Azure.Storage.DataMovement/src/CommitChunkHandler.cs

```diff
@@ -167,49 +174,69 @@ private async Task NotifyOfPendingStageChunkEvents()
                     {
                         // Add CommitBlockList task to the channel
                         await _queueCommitBlockTask().ConfigureAwait(false);
-                        _currentBytesSemaphore.Release();
-                        return;
                     }
                     else if (_bytesTransferred > _expectedLength)
                     {
-                        await _invokeFailedEventHandler(
-                                new Exception("Unexpected Error: Amount of bytes transferred exceeds expected length.")).ConfigureAwait(false);
-                        _currentBytesSemaphore.Release();
-                        return;
+                        throw Errors.MismatchLengthTransferred(
+                                expectedLength: _expectedLength,
+                                actualLength: _bytesTransferred);
                     }
                     _currentBytesSemaphore.Release();
                 }
             }
             catch (Exception ex)
             {
+                if (_currentBytesSemaphore.CurrentCount == 0)
```

Comment:
Can you explain this check a little? Why do we need to check this in the catch but not above in the happy path?

Reply:
In the happy path, it should be assumed that the semaphore should have been taken. If somehow we release, and the semaphore was never taken, we really should be throwing an exception in the happy path.

We check here, in the catch block, if the semaphore has been released here / or not taken yet, cause it's possible that the semaphore was not locked.

Reply:
Ok, makes sense. Thanks